src/libraries/Common/src/Interop/OSX/Interop.Network.Tls.cs

```diff
@@ -0,0 +1,153 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Net.Security;
+using System.Runtime.InteropServices;
+using System.Security.Authentication;
+using Microsoft.Win32.SafeHandles;
+
+internal static partial class Interop
+{
+    // TLS 1.3 specific Network Framework implementation for macOS
+    internal static partial class NetworkFramework
+    {
+        internal static partial class Tls
+        {
+            // Core TLS functions for Network Framework integration
+            [LibraryImport(Interop.Libraries.AppleNetworkNative, EntryPoint = "AppleNetNative_NwInit")]
+            [return: MarshalAs(UnmanagedType.I4)]
+            internal static unsafe partial bool Init(delegate* unmanaged<IntPtr, StatusUpdates, IntPtr, IntPtr, void> statusCallback,
+                                                   delegate* unmanaged<IntPtr, byte*, void**, int> readCallback,
+                                                   delegate* unmanaged<IntPtr, byte*, void**, int> writeCallback);
+
+            [LibraryImport(Interop.Libraries.AppleNetworkNative, EntryPoint = "AppleNetNative_NwCreateContext")]
+            internal static partial SafeNwHandle CreateContext([MarshalAs(UnmanagedType.I4)] bool isServer);
+
+            [LibraryImport(Interop.Libraries.AppleNetworkNative, EntryPoint = "AppleNetNative_NwSetTlsOptions", StringMarshalling = StringMarshalling.Utf8)]
+            private static partial int SetTlsOptions(SafeNwHandle connection, IntPtr gcHandle,
+                                                            string targetName, Span<byte> alpnBuffer, int alpnLength,
+                                                            SslProtocols minTlsProtocol, SslProtocols maxTlsProtocol);
+
+            internal static int SetTlsOptions(SafeNwHandle nwHandle, IntPtr gcHandle, string targetName, List<SslApplicationProtocol>? applicationProtocols, SslProtocols minTlsVersion, SslProtocols maxTlsVersion)
+            {
+                int alpnLength = GetAlpnProtocolListSerializedLength(applicationProtocols);
+                Span<byte> alpn = alpnLength <= 256 ? stackalloc byte[256].Slice(0, alpnLength) : new byte[alpnLength];
+                SerializeAlpnProtocolList(applicationProtocols, alpn);
+
+                return SetTlsOptions(nwHandle, gcHandle, targetName, alpn, alpnLength, minTlsVersion, maxTlsVersion);
+            }
+
+            [LibraryImport(Interop.Libraries.AppleNetworkNative, EntryPoint = "AppleNetNative_NwStartTlsHandshake")]
+            internal static partial int StartTlsHandshake(SafeNwHandle connection, IntPtr gcHandle);
```

Comment:
Is the `gcHandle` parameter somehow required to be a GC handle by the native code, or is this really just opaque state that's passed through?